onnx/onnx_backend.h

```diff
@@ -0,0 +1,914 @@
+ * backend.
+ *
+ * Value type: uint64_t.
+ * If the backend does not support FP32 computation, the value MUST be 0.
+ */
+#define ONNX_BACKEND_MACS_FP32 30
+
+/**
+ * Number of FP16 multiply-accumulate operations per second delivered by the
+ * backend.
+ *
+ * Value type: uint64_t.
+ * If the backend does not support FP16 computation, the value MUST be 0.
+ */
+#define ONNX_BACKEND_MACS_FP16 31
+
+/**
+ * Bandwidth, in bytes per second, of the global memory specific to the backend
+ * device.
+ *
+ * Value type: uint64_t.
+ */
+#define ONNX_BACKEND_MEMORY_BANDWIDTH 35
+
+/**
+ * Bandwidth, in bytes per second, of transferring data from cacheable
+ * CPU-allocated memory to the backend device.
+ *
+ * Value type: uint64_t.
+ */
+#define ONNX_BACKEND_CPU_MEMORY_READ_BANDWIDTH 36
+
+/**
+ * Bandwidth, in bytes per second, of transferring data to cacheable
+ * CPU-allocated memory from the backend device.
+ *
+ * Value type: uint64_t.
+ */
+#define ONNX_BACKEND_CPU_MEMORY_WRITE_BANDWIDTH 37
+
+/* Note: the data type values match ONNX TensorProto.DataType enum */
+#define ONNX_DATATYPE_UNDEFINED 0
+#define ONNX_DATATYPE_FLOAT16 10
+#define ONNX_DATATYPE_FLOAT32 1
+#define ONNX_DATATYPE_FLOAT64 11
+#define ONNX_DATATYPE_INT8 3
+#define ONNX_DATATYPE_INT16 5
+#define ONNX_DATATYPE_INT32 6
+#define ONNX_DATATYPE_INT64 7
+#define ONNX_DATATYPE_UINT8 2
+#define ONNX_DATATYPE_UINT16 4
+#define ONNX_DATATYPE_UINT32 12
+#define ONNX_DATATYPE_UINT64 13
+
+/** Cacheable CPU memory */
+#define ONNX_MEMORY_TYPE_CPU 0
+
+#define ONNX_TENSOR_DIMS_MAX 8
+
+/**
+ * Terminates the list of auxiliary backend initialization properties passed to
+ * onnxInitBackend.
+ */
+#define ONNX_BACKEND_PROPERTY_NONE 0
+
+typedef struct onnxTensorDescriptor {
+  /**
+   * Name of the blob corresponding to this tensor in the ONNX model. The name
+   * must exactly match the ValueInfoProto.name of one of the
+   * ModelProto.graph.input or ModelProto.graph.output
+   */
+  const char* name;
+  /**
+   * Data type of the elements in the tensor.
+   *
+   * Possible values:
+   *     ONNX_DATATYPE_FLOAT16
+   *     ONNX_DATATYPE_FLOAT32
+   *     ONNX_DATATYPE_INT8
+   *     ONNX_DATATYPE_INT16
+   *     ONNX_DATATYPE_INT32
+   *     ONNX_DATATYPE_UINT8
+   *     ONNX_DATATYPE_UINT16
+   *     ONNX_DATATYPE_UINT32
+   */
+  onnxEnum dataType;
+  /**
+   * Type of memory that stores the tensor.
+   *
+   * Possible values:
+   *     ONNX_MEMORY_TYPE_CPU
+   */
+  onnxEnum memoryType;
+  /**
+   * Number of dimensions in the tensor.
+   * Must be between 0 (for a scalar) and ONNX_TENSOR_DIMS_MAX.
+   */
+  uint32_t dimensions;
+  /**
+   * Dimensions of the tensor.
+   */
+  uint64_t shape[ONNX_TENSOR_DIMS_MAX];
+  /**
+   * Pointers to tensor data.
+   */
+  onnxPointer buffer;
+} onnxTensorDescriptor;
+
+/* Function pointer declarations for dynamic loading */
+typedef uint32_t(ONNX_ABI* onnxGetNumBackendsFunction)(void);
```

Comment:
This should be the API for hardware vendors, right? Each hardware vendor should be one backend.

Reply:
Hardware vendors would implement this API, and high-level applications would use this API. Deep learning frameworks could both use and implement this API.

An ONNX backend is a combination of software layer (`onnx-myvendor.dll` library) **and** a hardware device (e.g. `"NVIDIA Tesla P100"`). Thus, software library implementing this interface can expose multiple ONNX backends.

Reply:
Correct me if I'm wrong please. Are we expecting one backend (hardware accelerator or software vendor) wrapping against another bunch of backends?

Reply:
It is possible that a high-level framework (e.g. Caffe2, CNTK, or MXNet) implements that Backend API. Internally, the high-level framework can use also use Backend API to leverage hardware-specific backends exposed by vendors. Applications can distinguish between backends using a specific device and backends internally managing execution between multiple devices by doing `onnxGetBackendInfo` for `ONNX_BACKEND_DEVICE_TYPE`. The return value for such high-level backends spanning multiple hardware devices is `ONNX_DEVICE_TYPE_FRAMEWORK`, as opposed to `ONNX_DEVICE_TYPE_NPU`, `ONNX_DEVICE_TYPE_DSP`, `ONNX_DEVICE_TYPE_GPU`, `ONNX_DEVICE_TYPE_CPU`, or `ONNX_DEVICE_TYPE_FPGA` for backends corresponding to real hardware devices.

Reply:
A typical sequence of ONNX Backend API function calls on application side is documented in https://github.com/Maratyszcza/onnx/blob/ad150df9ef67843f9ae5649834f748d0691fd94c/docs/BackendAPI.md#how-to-use-onnx-backend-interface